### PR TITLE
 Add mocked supervised trainer test

### LINKED ISSUE #1751: Improve the supervised evaluator and trainer tests by using mocking
## 🚀 Feature

As it became clear in this PR: 1714. Some of the functionalities of both the supervised evaluator and trainer are not tested as we need to mock some of the underlying functions i.e. evaluate steps or engines.

I suggest mocking those objects and write test cases that test that those mocks were called in different evaluation scenarios. We already use mocks in some of our tests and we can start unify them across all tests that include complex execution scenarios.

### PR DESCRIPTION
**Pull request by Moh-Yakoub**
Fixes #1751

Description:

I've added a mocked version of the trainer tests to be able to test that we are using the correct trainer step
Check list:

New tests are added (if a new feature is added)
New doc strings: description and/or example code are in RST format
Documentation is updated (if required)

### CODE REVIEW

**Review by vfdev-5 — APPROVED** [review]
Thanks @Moh-Yakoub ! LGTM

**Comment by vfdev-5** [comment]
@Moh-Yakoub please resolve the conflict due to another merged PR and we are good to land it.

**Comment by Moh-Yakoub** [comment]
> @Moh-Yakoub please resolve the conflict due to another merged PR and we are good to land it. Thanks

@vfdev-5 Conflicts were resolved. I think this PR is ready

File: tests/ignite/engine/test_create_supervised.py [hunk 1]
```diff
@@ -79,6 +84,74 @@ def _test_create_supervised_trainer(
...
+                        elif mode == "tpu":
+                            assert training_step_tpu_mock.called
+                        else:
+                            assert training_step_mock.called
+
+                    else:
+                        if LooseVersion(torch.__version__) >= LooseVersion("1.7.0"):
+                            # This is broken in 1.6.0 but will be probably fixed with 1.7.0
+                            with pytest.raises(RuntimeError, match=r"is on CPU, but expected them to be on GPU"):
+                                trainer.run(data)
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:152]: @Moh-Yakoub this should be removed otherwise it will break our CI, see #1800

**Review by vfdev-5 — APPROVED** [review]

**Pull request merged**